File: solutions/rust/forth/1/src/handlers/stack_helpers.rs
use crate::Error;
use crate::models::context::Context;
use crate::utilities::list_extensions::{require_at_least_one_value, require_at_least_two_values};

pub fn duplicate_last(context: &mut Context) -> Result<Context, Error> {
	context.with_stack(|stack| {
		let values = require_at_least_one_value(stack)?;
		let last = values.last().copied().unwrap_or_default();
		stack.push(last);
		Ok(())
	})
}

pub fn drop_last(context: &mut Context) -> Result<Context, Error> {
	context.with_stack(|stack| {
		let _values = require_at_least_one_value(stack)?;
		stack.pop();
		Ok(())
	})
}

pub fn swap_last(context: &mut Context) -> Result<Context, Error> {
	context.with_stack(|stack| {
		let values = require_at_least_two_values(stack)?;
		let last_value = values.pop();
		let penultimate_value = stack.pop().unwrap();
		stack.push(last_value.unwrap());
		stack.push(penultimate_value);
		Ok(())
	})
}

pub fn penultimate_value_copy(context: &mut Context) -> Result<Context, Error> {
	context.with_stack(|stack| {
		let values = require_at_least_two_values(stack)?;
		if let Some(&penultimate_value) = values.get(values.len() - 2) {
			stack.push(penultimate_value);
		}
		Ok(())
	})
}

File: solutions/ocaml/forth/6/stack_helpers.ml
let duplicate_last (context : Context.t) : Context.t =
  Context.with_stack context (fun stack ->
    let last =
      !stack
      |> List_helpers.require_at_least_one_value
      |> List_helpers.get_last_element
    in
    stack := !stack @ [last]
  )

let drop_last (context : Context.t) : Context.t =
  Context.with_stack context (fun stack ->
    let _ = List_helpers.require_at_least_one_value !stack in
    let _ = List_helpers.pop stack in
    ()
  )

let swap_last (context : Context.t) : Context.t =
  Context.with_stack context (fun stack ->
    let _ = List_helpers.require_at_least_two_values !stack in
    let last_value = List_helpers.pop stack in
    let penultimate_value = List_helpers.pop stack in
    stack := !stack @ [last_value; penultimate_value]
  )

let penultimate_value_copy (context : Context.t) : Context.t =
  Context.with_stack context (fun stack ->
    let penultimate_value =
      !stack
      |> List_helpers.require_at_least_two_values
      |> fun lst -> List.nth lst (List.length lst - 2)
    in
    stack := !stack @ [penultimate_value]
  )
